pkg/controllers/clusterresourceplacementeviction/controller.go

```diff
@@ -0,0 +1,257 @@
+		}
+		isCRPPresent = false
+	}
+	if !isCRPPresent {
+		klog.V(2).InfoS(evictionInvalidMissingCRP, "clusterResourcePlacementEviction", evictionName, "clusterResourcePlacement", eviction.Spec.PlacementName)
+		markEvictionInvalid(&eviction, evictionInvalidMissingCRP)
+		return runtime.Result{}, r.updateEvictionStatus(ctx, &eviction)
+	}
+
+	var crbList placementv1beta1.ClusterResourceBindingList
+	if err := r.Client.List(ctx, &crbList, client.MatchingLabels{placementv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
+		return runtime.Result{}, err
+	}
+
+	var evictionTargetBinding *placementv1beta1.ClusterResourceBinding
+	for i := range crbList.Items {
+		if crbList.Items[i].Spec.TargetCluster == eviction.Spec.ClusterName {
+			if evictionTargetBinding == nil {
+				evictionTargetBinding = &crbList.Items[i]
+			} else {
+				klog.V(2).InfoS(evictionInvalidMultipleCRB, "clusterResourcePlacementEviction", evictionName, "clusterResourcePlacement", eviction.Spec.PlacementName)
+				markEvictionInvalid(&eviction, evictionInvalidMultipleCRB)
+				return runtime.Result{}, r.updateEvictionStatus(ctx, &eviction)
+			}
+		}
+	}
+	if evictionTargetBinding == nil {
+		klog.V(2).InfoS(evictionInvalidMissingCRB, "clusterResourcePlacementEviction", evictionName, "clusterName", eviction.Spec.ClusterName)
+		markEvictionInvalid(&eviction, evictionInvalidMissingCRB)
+		return runtime.Result{}, r.updateEvictionStatus(ctx, &eviction)
+	}
+
+	markEvictionValid(&eviction)
+	isDBPresent := true
+	var db placementv1alpha1.ClusterResourcePlacementDisruptionBudget
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: crp.Name}, &db); err != nil {
+		if !errors.IsNotFound(err) {
+			return runtime.Result{}, err
+		}
+		isDBPresent = false
+	}
+
+	if !isDBPresent {
+		if err := r.deleteClusterResourceBinding(ctx, evictionTargetBinding); err != nil {
+			return runtime.Result{}, err
+		}
+		markEvictionExecuted(&eviction, evictionAllowedNoPDB)
+		return runtime.Result{}, r.updateEvictionStatus(ctx, &eviction)
+	}
+
+	var desiredBindings int
+	switch crp.Spec.Policy.PlacementType {
+	case placementv1beta1.PickAllPlacementType:
+		desiredBindings = len(crbList.Items)
+	case placementv1beta1.PickNPlacementType:
+		desiredBindings = int(*crp.Spec.Policy.NumberOfClusters)
+	case placementv1beta1.PickFixedPlacementType:
+		desiredBindings = len(crp.Spec.Policy.ClusterNames)
+	}
+
+	totalBindings := len(crbList.Items)
+	allowed, disruptionsAllowed, availableBindings := isEvictionAllowed(desiredBindings, crbList.Items, db)
+	if allowed {
+		if err := r.deleteClusterResourceBinding(ctx, evictionTargetBinding); err != nil {
+			return runtime.Result{}, err
+		}
+		markEvictionExecuted(&eviction, fmt.Sprintf(evictionAllowedPDBSpecified, disruptionsAllowed, availableBindings, desiredBindings, totalBindings))
+	} else {
+		markEvictionNotExecuted(&eviction, fmt.Sprintf(evictionBlockedPDBSpecified, disruptionsAllowed, availableBindings, desiredBindings, totalBindings))
```

Comment:
this will stop the eviction forever as per line 71?

Reply:
disruptionsAllowed is always <= 0 when it is blocked, that number seems not useful

Reply:
The following states are terminal states for an eviction, so we block reconcile.
- invalid eviction
- executed eviction
- not executed eviction

The user needs to retry by creating another eviction